src/core/reaction_ensemble.cpp

```diff
@@ -0,0 +1,2061 @@
+		factorial_expr=factorial_expr*factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i(N_i0,nu_i); //zeta = 1 (see smith paper) since we only perform one reaction at one call of the function
+	}
+	//factorial contribution of products
+	for(int i=0;i<current_reaction->len_product_types;i++) {
+		int nu_i=current_reaction->product_coefficients[i];
+		int N_i0= old_particle_numbers[find_index_of_type(current_reaction->product_types[i])];
+		factorial_expr=factorial_expr*factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i(N_i0,nu_i); //zeta = 1 (see smith paper) since we only perform one reaction at one call of the function
+	}
+	return factorial_expr;
+}
+
+int generic_oneway_reaction(int reaction_id){
+	double volume = current_reaction_system.volume;
+	single_reaction* current_reaction=current_reaction_system.reactions[reaction_id];
+	
+	//generic one way reaction
+	//A+B+...+G +... --> K+...X + Z +...
+	//you need to use 2A --> B instead of A+A --> B since in the last case you assume distinctness of the particles, however both ways to describe the reaction are equivalent in the thermodynamic limit
+	//further it is crucial for the function in which order you provide the reactant and product types since particles will be replaced correspondingly!
+	
+	if (all_reactant_particles_exist(reaction_id) ==false ) {
+		//makes sure, no incomplete reaction is performed -> only need to consider rollback of complete reactions
+		return 0;
+	}
+	
+	//calculate potential energy
+	double E_pot_old=calculate_current_potential_energy_of_system(0); //only consider potential energy since we assume that the kinetic part drops out in the process of calculating ensemble averages (kinetic part may be seperated and crossed out)
+	
+	//find reacting molecules in reactants and save their properties for later recreation if step is not accepted
+	//do reaction
+	//save old particle_numbers
+	int* old_particle_numbers=(int*) calloc(1,sizeof(int) *(current_reaction_system.nr_different_types));
+	for(int type_index=0;type_index<current_reaction_system.nr_different_types;type_index++){
+		number_of_particles_with_type(current_reaction_system.type_index[type_index], &(old_particle_numbers[type_index])); // here could be optimized by not going over all types but only the types that occur in the reaction
+	}
+
+	int* p_ids_created_particles =NULL;
+	int len_p_ids_created_particles=0;
+	double* hidden_particles_properties=NULL;
+	int len_hidden_particles_properties=0;
+	double* changed_particles_properties=NULL;
+	int len_changed_particles_properties=0;
+	int number_of_saved_properties=3; //save p_id, charge and type of the reactant particle, only thing we need to hide the particle and recover it
+	make_reaction_attempt(current_reaction, &changed_particles_properties, &len_changed_particles_properties, &p_ids_created_particles, &len_p_ids_created_particles, &hidden_particles_properties, &len_hidden_particles_properties);
```

Comment:
If you make a new type ReactionAttempt, you can make this much less messy. Also, since as far as I can see this is used only locally it should be easy to replace the pointer mess with vectors.